image-service/src/main/kotlin/com/instagram_clone/image_service/service/ImageMetaServiceMongoImpl.kt

```diff
@@ -50,6 +53,120 @@ class ImageMetaServiceMongoImpl(private val client: MongoClient) : ImageMetaServ
     return promise.future()
   }
 
+  override fun likeImage(
+    imageId: String,
+    userId: String
+  ): Future<Nothing> = getImageMeta(imageId)
+    .compose { meta ->
+      if (meta == null) {
+        throw NotFoundException("No image found with id $imageId")
+      }
+      addLike(imageId, userId)
+    }
+    .compose { newLike ->
+      // Increase like counter only if the like is "new"
+      if (newLike) {
+        increaseLikeCounter(imageId, 1)
+      } else {
+        Future.succeededFuture()
+      }
+    }
+
+  override fun unlikeImage(
+    imageId: String,
+    userId: String
+  ): Future<Nothing> = getImageMeta(imageId)
+    .compose { meta ->
+      if (meta == null) {
+        throw NotFoundException("No image found with id $imageId")
+      }
+      removeLike(imageId, userId)
+    }
+    .compose { removed ->
+      if (removed) {
+        increaseLikeCounter(imageId, -1)
+      } else {
+        Future.succeededFuture()
+      }
+    }
+
+  /**
+   * Remove the like of [imageId] from [userId]. Return boolean indicating whether
+   * the like actually existed or not (success in any case).
+   */
+  private fun removeLike(imageId: String, userId: String): Future<Boolean> {
+    val promise = Promise.promise<Boolean>()
+
+    val query = json {
+      obj("imageId" to imageId, "userId" to userId)
+    }
+
+    client.findOneAndDelete(config.likesCollection, query) {
+      if (it.succeeded()) {
+        promise.complete(true)
+      } else {
+        promise.fail(it.cause())
+      }
+    }
+    return promise.future()
+  }
+
+  /**
+   * Add a like if it doesn't exist based on [imageId] and [userId]. Return
+   * boolean describing whether new like was inserted or not (already existing).
+   */
+  private fun addLike(imageId: String, userId: String): Future<Boolean> {
+    val promise = Promise.promise<Boolean>()
+    val query = json {
+      obj("imageId" to imageId, "userId" to userId)
+    }
+    val update = json {
+      obj("\$set" to query)
+    }
+    client.updateCollectionWithOptions(
+      config.likesCollection,
+      query,
+      update,
+      UpdateOptions(true)
+    ) {
+      if (it.succeeded()) {
+        val r = it.result()
+        if (r.docUpsertedId != null && r.docMatched == 0L) {
+          promise.complete(true)
+        } else {
+          promise.complete(false);
+        }
+      } else {
+        promise.fail(it.cause())
+      }
+    }
+    return promise.future()
+  }
+
+  /**
+   * Increase the like counter based on [imageId], for the amount of [amount]. Please
+   * note that negative [amount] values will decrease the "likes" counter.
+   */
+  private fun increaseLikeCounter(imageId: String, amount: Long): Future<Nothing> {
```

Comment:
Is there a case when the amount would be different than -1 or 1. Maybe it would better to just pass a `increase: bool` or something.

Reply:
Yeah you're probs right, it's always -1 or 1

Reply:
This is now fixed